packages/react-graphql/src/hooks/useQuery.ts

```diff
@@ -0,0 +1,202 @@
+    client: overrideClient,
+    context,
+    variables,
+    fetchPolicy,
+    errorPolicy,
+  } = options;
+
+  const [responseId, setResponseId] = useState(0);
+  const client = useApolloClient(overrideClient);
+  const query = useGraphQLDocument(queryOrComponent);
+
+  const watchQueryOptions: WatchQueryOptions<Variables> = useMemo(
+    () =>
+      compact({
+        context,
+        errorPolicy,
+        fetchPolicy,
+        notifyOnNetworkStatusChange,
+        pollInterval,
+        query,
+        variables,
+      }),
+    [
+      query,
+      pollInterval,
+      notifyOnNetworkStatusChange,
+      context && objectToKey(context),
+      variables && objectToKey(variables),
+      fetchPolicy,
+      errorPolicy,
+    ],
+  );
+
+  const observableQuery = useMemo(
+    () => {
+      if (!query) {
+        return;
+      }
+
+      // eslint-disable-next-line consistent-return
+      return getCachedObservableQuery<Data, Variables>(
+        client,
+        watchQueryOptions,
+      );
+    },
+    [query, client, watchQueryOptions],
+  );
+
+  useEffect(
+    () => {
+      if (skip || !observableQuery) {
+        return;
+      }
+
+      const invalidateCurrentResult = () => {
+        // A hack to get rid React warnings during tests. The default
+        // implementation of `actHack` just invokes the callback immediately.
+        // In tests, it's replaced with `act` from react-testing-library.
+        // A better solution welcome.
+        actHack(() => {
+          setResponseId(x => x + 1);
+        });
+      };
+      const subscription = observableQuery.subscribe(
+        invalidateCurrentResult,
+        invalidateCurrentResult,
+      );
+
+      invalidateCachedObservableQuery(client, watchQueryOptions);
+
+      // eslint-disable-next-line consistent-return
+      return () => {
+        subscription.unsubscribe();
+      };
+    },
+    [skip, observableQuery],
+  );
+
+  const currentResult = useMemo<QueryHookResult<Data, Variables> | {}>(
+    () => {
+      if (!observableQuery) {
+        return {};
+      }
+
+      const result = observableQuery.currentResult();
+
+      // return the old result data when there is an error
+      let data = result.data as Data;
+      if (result.error || result.errors) {
+        data = {
+          ...(result.data as Object),
+          ...((observableQuery.getLastResult() || {}).data as Object),
+        } as Data;
+      }
+
+      return {
+        data,
+        error:
+          result.errors && result.errors.length > 0
+            ? new ApolloError({graphQLErrors: result.errors})
+            : result.error,
+        loading: result.loading,
+        networkStatus: result.networkStatus,
+        partial: result.partial,
+      };
+    },
+    [skip, responseId, observableQuery],
+  );
+
+  if (!observableQuery) {
+    return {
+      loading: true,
+      networkStatus: undefined,
+    };
+  }
+
+  return {
+    ...getDefaultQueryResult(client, observableQuery),
```

Comment:
_carry over things that's not resolved yet but lost due to moving files_

https://github.com/Shopify/quilt/pull/642#discussion_r273722289
> Feels to me like we should do this in the `useMemo` where we create currentResult instead?